very simple code using np.abs, numpy.linalg.norm fails to compile

This code fails to compile:

https://gist.github.com/nbecker/d990c020a6f05d25b302b745f05c1588

Problem seems to be
/usr/bin/../lib/gcc/x86_64-redhat-linux/10/../../../../include/c++/10/bits/stl_algobase.h:380:20: error: assigning to 'double' from incompatible type 'const std::complex<double>'

Both np.abs and norm return scalars, so I don't see anything that should be std::complex<double> assigning to double.

--- a/pythran/pythonic/include/numpy/linalg/norm.hpp
+++ b/pythran/pythonic/include/numpy/linalg/norm.hpp
@@ -1,9 +1,8 @@
 #ifndef PYTHONIC_INCLUDE_NUMPY_LINALG_NORM_HPP
 #define PYTHONIC_INCLUDE_NUMPY_LINALG_NORM_HPP
 
-#include "pythonic/include/numpy/abs.hpp"
 #include "pythonic/include/numpy/sqrt.hpp"
-#include "pythonic/include/numpy/square.hpp"
+#include "pythonic/include/builtins/pythran/abssqr.hpp"
 #include "pythonic/include/numpy/sum.hpp"
 #include "pythonic/include/numpy/asfarray.hpp"
 
@@ -15,11 +14,10 @@ namespace numpy
     template <class Array>
     auto norm(Array &&array, types::none_type ord = {},
               types::none_type axis = {})
-        -> decltype(pythonic::numpy::functor::sqrt{}(
-            pythonic::numpy::functor::sum{}(pythonic::numpy::functor::square{}(
-                pythonic::numpy::functor::abs{}(
-                    pythonic::numpy::functor::asfarray{}(
-                        std::forward<Array>(array)))))));
+        -> decltype(
+            pythonic::numpy::functor::sqrt{}(pythonic::numpy::functor::sum{}(
+                pythonic::builtins::pythran::functor::abssqr{}(
+                    std::forward<Array>(array)))));
 
     template <class Array>
     using norm_dtype_t = typename std::conditional<

--- a/pythran/pythonic/numpy/linalg/norm.hpp
+++ b/pythran/pythonic/numpy/linalg/norm.hpp
@@ -10,7 +10,7 @@
 #include "pythonic/numpy/power.hpp"
 #include "pythonic/numpy/real.hpp"
 #include "pythonic/numpy/sqrt.hpp"
-#include "pythonic/numpy/square.hpp"
+#include "pythonic/builtins/pythran/abssqr.hpp"
 #include "pythonic/numpy/sum.hpp"
 #include "pythonic/builtins/NotImplementedError.hpp"
 PYTHONIC_NS_BEGIN
@@ -20,16 +20,14 @@ namespace numpy
   {
     template <class Array>
     auto norm(Array &&array, types::none_type ord, types::none_type axis)
-        -> decltype(pythonic::numpy::functor::sqrt{}(
-            pythonic::numpy::functor::sum{}(pythonic::numpy::functor::square{}(
-                pythonic::numpy::functor::abs{}(
-                    pythonic::numpy::functor::asfarray{}(
-                        std::forward<Array>(array)))))))
+        -> decltype(
+            pythonic::numpy::functor::sqrt{}(pythonic::numpy::functor::sum{}(
+                pythonic::builtins::pythran::functor::abssqr{}(
+                    std::forward<Array>(array)))))
     {
       return pythonic::numpy::functor::sqrt{}(pythonic::numpy::functor::sum{}(
-          pythonic::numpy::functor::square{}(pythonic::numpy::functor::abs{}(
-              pythonic::numpy::functor::asfarray{}(
-                  std::forward<Array>(array))))));
+          pythonic::builtins::pythran::functor::abssqr{}(
+              std::forward<Array>(array))));
     }
 
     template <class Array>

--- a/pythran/tests/test_numpy_linalg.py
+++ b/pythran/tests/test_numpy_linalg.py
@@ -31,6 +31,11 @@ class TestNumpyLinalg(TestEnv):
     def test_linalg_norm7(self):
         self.run_test("def linalg_norm7(x): from numpy.linalg import norm ; return norm(x)", numpy.arange(6).reshape(2,3), linalg_norm7=[NDArray[int,:,:]])
 
+    def test_linalg_norm8(self):
+        self.run_test("def linalg_norm8(x): from numpy.linalg import norm ; return norm(x)",
+                      numpy.arange(6).reshape(2,3) * 1j + 1,
+                      linalg_norm8=[NDArray[complex,:,:]])
+
     def test_linalg_norm_pydoc(self):
         self.run_test('''
             def linalg_norm_pydoc(x):

